File: win-prog/ProgWinCS/Images and Bitmaps/Thumbnail/Thumbnail.cs
//----------------------------------------
// Thumbnail.cs © 2001 by Charles Petzold
//----------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class Thumbnail: PrintableForm
{
     const int iSquare = 64;
     Image imageThumbnail;

     public new static void Main()
     {
          Application.Run(new Thumbnail());
     }
     public Thumbnail()
     {
          Text = "Thumbnail";

          Image image = Image.FromFile("..\\..\\..\\Apollo11FullColor.jpg");

          int cxThumbnail, cyThumbnail;

          if (image.Width > image.Height)
          {
               cxThumbnail = iSquare;
               cyThumbnail = iSquare * image.Height / image.Width;
          }
          else
          {
               cyThumbnail = iSquare;
               cxThumbnail = iSquare * image.Width / image.Height;
          }
          imageThumbnail = image.GetThumbnailImage(cxThumbnail, cyThumbnail,
                                                   null, (IntPtr) 0);
     }
     protected override void DoPage(Graphics grfx, Color clr, int cx, int cy)
     {
          for (int y = 0; y < cy; y += iSquare)
          for (int x = 0; x < cx; x += iSquare)
               grfx.DrawImage(imageThumbnail, 
                              x + (iSquare - imageThumbnail.Width) / 2,
                              y + (iSquare - imageThumbnail.Height) / 2,
							  imageThumbnail.Width, imageThumbnail.Height);
     }
} 

File: win-prog/ProgWinVB/Images and Bitmaps/Thumbnail/Thumbnail.vb
'------------------------------------------
' Thumbnail.vb (c) 2002 by Charles Petzold
'------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class Thumbnail
    Inherits PrintableForm

    Const iSquare As Integer = 64
    Private imgThumbnail As Image

    Shared Shadows Sub Main()
        Application.Run(new Thumbnail())
    End Sub

    Sub New()
        Text = "Thumbnail"
        Dim img As Image = Image.FromFile("..\..\Apollo11FullColor.jpg")
        Dim cxThumbnail, cyThumbnail As Integer

        If img.Width > img.Height Then
            cxThumbnail = iSquare
            cyThumbnail = iSquare * img.Height \ img.Width
        Else
            cyThumbnail = iSquare
            cxThumbnail = iSquare * img.Width \ img.Height
        End If
        imgThumbnail = img.GetThumbnailImage(cxThumbnail, cyThumbnail, _
                                      Nothing, IntPtr.Zero)
    End Sub

    Protected Overrides Sub DoPage(ByVal grfx As Graphics, _
            ByVal clr As Color, ByVal cx As Integer, ByVal cy As Integer)
        Dim x, y As Integer

        For y = 0 To cy Step iSquare
            For x = 0 To cx Step iSquare
                grfx.DrawImage(imgThumbnail, _
                        x + (iSquare - imgThumbnail.Width) \ 2, _
                        y + (iSquare - imgThumbnail.Height) \ 2, _
                        imgThumbnail.Width, imgThumbnail.Height)
            Next x
        Next y
    End Sub
End Class
